src/opam.nix

```diff
@@ -480,4 +482,120 @@ in rec {
+    # This is O(n^2). We could try and improve this by sorting the list on name. But n is small.
+    let op = srcs: newSrc:
+      # Find if two packages come from the same dev-repo.
+      # Note we are assuming no dev-repos will have different names here, but we also assume
+      # this later when we will symlink in the duniverse directory based on this name.
+      let duplicateSrc = findSingle (src: src.name == newSrc.name) null "multiple" srcs; in
+      # Multiple duplicates should never be found as we deduplicate on every new element.
+      assert duplicateSrc != "multiple";
+      if duplicateSrc == null then srcs ++ [ newSrc ]
+      # > If packages from the same repo were resolved to different URLs, we need to pick
+      # > a single one. Here we decided to go with the one associated with the package
+      # > that has the higher version. We need a better long term solution as this won't
+      # > play nicely with pins for instance.
+      # > The best solution here would be to use source trimming, so we can pull each individual
+      # > package to its own directory and strip out all the unrelated source code but we would
+      # > need dune to provide that feature.
+      # See [opam-monorepo](https://github.com/tarides/opam-monorepo/blob/9262e7f71d749520b7e046fbd90a4732a43866e9/lib/duniverse.ml#L143-L157)
+      else if duplicateSrc.version >= newSrc.version then srcs
+      else (remove duplicateSrc srcs) ++ [ newSrc ];
+    in foldl' op [ ] srcs;
+
+  mkMonorepo = srcs:
+    let
+      # derivation that fetches the source
+      mkSrc = { name, version, src }:
+        bootstrapPackages.pkgsBuildBuild.stdenv.mkDerivation ({
+          inherit name version src;
+          phases = [ "unpackPhase" "installPhase" ];
+          installPhase = ''
+            mkdir $out
+            cp -R . $out
+          '';
+        });
+    in
+    listToAttrs (map (src: nameValuePair src.name (mkSrc src)) srcs);
+
+  queryToMonorepo = { repos ? [ opamOverlays opamRepository ], resolveArgs ? { }
+      , filterPkgs ? [ ] }:
+    query:
+    pipe query [
+      # pass monorepo = 1 to pick up dependencies marked with {?monorepo}
+      (opamList (joinRepos repos) (resolveArgs // { env.monorepo=1; }))
+      opamListToQuery
+      (queryToDefs repos)
+      (defsToSrcs filterPkgs)
+      deduplicateSrcs
+      mkMonorepo
+    ];
+
+  opamRepositoryModuloOverlay =
+    let
+      overlayPkgPaths = pipe (opamOverlays + "/packages") [
+        readDir
+        (filterAttrs (name: type: type == "directory"))
+        attrNames
+        (map (pkgname: "packages/${pkgname}"))
+      ];
+    in path {
+      name = "opam-repository-modulo-overlay";
+      path = opamRepository;
+      # filter out if false, keep if true
+      filter = path: type: type != "directory" ||
+        # O(n^2)...
+        # if there isn't a match for an package in opam-overlay
+        ! elem true (map (overlayPkgPath: hasInfix overlayPkgPath path) overlayPkgPaths);
```

Comment:
```suggestion
        ! any (overlayPkgPath: hasInfix overlayPkgPath path) overlayPkgPaths;
```

Reply:
Also, `hasInfix` might be a bit flaky here, and quite slow.

On a more general note, maybe there's a way to just make a custom `joinRepos` that would overwrite entire `package`'s instead of just package versions. That would remove the slowdown, but will probably require modifying the `symlinkJoin` from nixpkgs.

Reply:
That just made me realise that `joinRepos` is not quite correct ATM, since it doesn't actually do the version overriding properly. I think I'll fix it and implement this at the same time.

Reply:
That sounds like a good idea, but I should warn you I think it will suffer from the same issues I described in the other open comment.

Essentially there's two possibilities when there's a package in opam-overlay with a newer in opam-repository without an overlay. One, an overlay hasn't been written yet. Two, the new version now supports dune and an overlay isn't required. In the former we want to remove the new version from opam-repository, in the latter we don't.

I don't think there's a way to differentiate these cases without parsing the new version's opam file to check for a dune dependency which suggests (although I guess doesn't guarantee, but it seems to work fine for opam-monorepo) dune support).

I'm ironing out the bugs on a function to do this now. It's not fast, but I think it'll work. I'll be offline for the next couple of hours but will push what I have now.

Reply:
Okay it seems to be working! (Although it's still building... I might lose network connection soon so pushing preemptively)

Reply:
Ahh, ok, sorry for not reading thoroughly. I still made the `joinRepos` a bit better (and quite a lot faster actually) by patching `lndir` (which comes from... Xorg of all places?)

https://github.com/tweag/opam-nix/pull/20